Adopt ASCIILiteral in TextEncoding code

cdumez opened the pull request:
#### e59e03e1df5b9d3dce970b9590767ceaabb9da47
<pre>
Adopt ASCIILiteral in TextEncoding code
<a href="https://bugs.webkit.org/show_bug.cgi?id=264988">https://bugs.webkit.org/show_bug.cgi?id=264988</a>

Reviewed by Darin Adler.

Adopt ASCIILiteral in TextEncoding code instead of `const char*` to
clarify the lifetime of these strings.

* Source/WTF/wtf/HashTraits.h:
(WTF::HashTraits&lt;ASCIILiteral&gt;::emptyValue):
(WTF::HashTraits&lt;ASCIILiteral&gt;::constructDeletedValue):
(WTF::HashTraits&lt;ASCIILiteral&gt;::isDeletedValue):
* Source/WTF/wtf/text/ASCIILiteral.h:
(WTF::ASCIILiteral::hash const):
(WTF::ASCIILiteralHash::hash):
(WTF::ASCIILiteralHash::equal):
* Source/WebCore/PAL/pal/text/TextCodec.h:
* Source/WebCore/PAL/pal/text/TextCodecCJK.cpp:
(PAL::TextCodecCJK::registerEncodingNames):
(PAL::TextCodecCJK::registerCodecs):
* Source/WebCore/PAL/pal/text/TextCodecICU.cpp:
(PAL::TextCodecICU::registerCodecs):
(PAL::TextCodecICU::TextCodecICU):
* Source/WebCore/PAL/pal/text/TextCodecICU.h:
* Source/WebCore/PAL/pal/text/TextCodecLatin1.cpp:
(PAL::TextCodecLatin1::registerEncodingNames):
(PAL::TextCodecLatin1::registerCodecs):
* Source/WebCore/PAL/pal/text/TextCodecReplacement.cpp:
(PAL::TextCodecReplacement::registerEncodingNames):
(PAL::TextCodecReplacement::registerCodecs):
* Source/WebCore/PAL/pal/text/TextCodecSingleByte.cpp:
(PAL::TextCodecSingleByte::registerEncodingNames):
(PAL::TextCodecSingleByte::registerCodecs):
* Source/WebCore/PAL/pal/text/TextCodecUTF16.cpp:
(PAL::TextCodecUTF16::registerEncodingNames):
(PAL::TextCodecUTF16::registerCodecs):
* Source/WebCore/PAL/pal/text/TextCodecUTF8.cpp:
(PAL::TextCodecUTF8::registerEncodingNames):
(PAL::TextCodecUTF8::registerCodecs):
* Source/WebCore/PAL/pal/text/TextCodecUserDefined.cpp:
(PAL::TextCodecUserDefined::registerEncodingNames):
(PAL::TextCodecUserDefined::registerCodecs):
* Source/WebCore/PAL/pal/text/TextEncoding.cpp:
(PAL::TextEncoding::decode const):
(PAL::TextEncoding::encode const):
(PAL::TextEncoding::domName const):
(PAL::TextEncoding::usesVisualOrdering const):
* Source/WebCore/PAL/pal/text/TextEncoding.h:
(PAL::TextEncoding::isValid const):
(PAL::TextEncoding::name const):
* Source/WebCore/PAL/pal/text/TextEncodingRegistry.cpp:
(PAL::TextEncodingNameHash::hash):
(PAL::HashTranslatorTextEncodingName::hash):
(PAL::HashTranslatorTextEncodingName::equal):
(PAL::isUndesiredAlias):
(PAL::WTF_REQUIRES_LOCK):
(PAL::isJapaneseEncoding):
(PAL::shouldShowBackslashAsCurrencySymbolIn):
(PAL::newTextCodec):
(PAL::atomCanonicalTextEncodingName):
* Source/WebCore/PAL/pal/text/TextEncodingRegistry.h:
* Source/WebCore/dom/DocumentInlines.h:
(WebCore::Document::encoding const):

Canonical link: <a href="https://commits.webkit.org/270903@main">https://commits.webkit.org/270903@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57f4979ea240a9950846a8d52b762be21d3c87d7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26752 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5367 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27993 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28964 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24460 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27197 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7198 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2763 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24366 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27014 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4188 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22966 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3678 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/3720 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23959 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29449 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/23306 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24402 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24358 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29983 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/25945 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3757 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1951 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27886 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5206 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/33393 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6430 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4226 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7214 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4109 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/PAL/pal/text/TextCodecICU.cpp:
```diff
@@ -129,16 +129,16 @@ void TextCodecICU::registerCodecs(TextCodecRegistrar registrar)
         }
         registrar(name, [name, canonicalConverterName] {
-            return makeUnique<TextCodecICU>(name, canonicalConverterName);
+            return makeUnique<TextCodecICU>(name, ASCIILiteral::fromLiteralUnsafe(canonicalConverterName));
```
Add a comment explaining why the unsafe named function is safe to call here?

darinadler commented on Source/WebCore/PAL/pal/text/TextEncoding.cpp:
```diff
@@ -62,15 +62,15 @@ TextEncoding::TextEncoding(const String& name)
 String TextEncoding::decode(const char* data, size_t length, bool stopOnError, bool& sawError) const
 {
-    if (!m_name)
+    if (m_name.isEmpty())
```
This is a semantic change. The old code checks for null and the new code checks for empty. Just want to make sure you did that intentionally.

cdumez replied:
Ouch, this wasn't at all intentional. I am glad you caught it, thank you.

darinadler commented on Source/WebCore/PAL/pal/text/TextEncoding.cpp:
```diff
@@ -93,9 +93,9 @@ const char* TextEncoding::domName() const
-    static const char* const a = atomCanonicalTextEncodingName("windows-949");
-    if (m_name == a)
-        return "EUC-KR";
+    static const ASCIILiteral windows949 = atomCanonicalTextEncodingName("windows-949");
```
`constexpr`?

cdumez replied:
This doesn't build:
```
constexpr variable 'windows949' must be initialized by a constant expression
```

darinadler commented on Source/WebCore/PAL/pal/text/TextEncoding.cpp:
```diff
@@ -104,7 +104,7 @@ bool TextEncoding::usesVisualOrdering() const
         return false;
 
-    static const char* const a = atomCanonicalTextEncodingName("ISO-8859-8");
+    static const ASCIILiteral a = atomCanonicalTextEncodingName("ISO-8859-8");
```
`constexpr`?

cdumez replied:
Doesn't build with constexpr.

darinadler commented on Source/WebCore/PAL/pal/text/TextEncodingRegistry.cpp:
```diff
@@ -81,77 +82,89 @@ struct TextEncodingNameHash {
-    for (const char* p = alias; *p; ++p) {
-        if (*p == ',')
+    for (unsigned i = 0; i < alias.length(); ++i) {
+        if (alias.characterAt(i) == ',')
```
Surprised we have to write out a loop just to check for the presence of a character. Seems like `strchr` and `contains`.

cdumez replied:
Switching to `strchr()`.

darinadler commented on Source/WebCore/dom/DocumentInlines.h:
```diff
@@ -51,7 +51,11 @@ inline PAL::TextEncoding Document::textEncoding() const
 }
 
-inline AtomString Document::encoding() const { return AtomString::fromLatin1(textEncoding().domName()); }
+inline AtomString Document::encoding() const
```
Is this function really hot enough that it should be inlined?

cdumez replied:
Likely not.

darinadler commented on Source/WebCore/dom/DocumentInlines.h:
```diff
@@ -51,7 +51,11 @@ inline PAL::TextEncoding Document::textEncoding() const
 }
 
-inline AtomString Document::encoding() const { return AtomString::fromLatin1(textEncoding().domName()); }
+inline AtomString Document::encoding() const
```
Is this function really hot enough that it should be inline?

cdumez committed:
Adopt ASCIILiteral in TextEncoding code
https://bugs.webkit.org/show_bug.cgi?id=264988

Reviewed by Darin Adler.

Adopt ASCIILiteral in TextEncoding code instead of `const char*` to
clarify the lifetime of these strings.

* Source/WTF/wtf/HashTraits.h:
(WTF::HashTraits<ASCIILiteral>::emptyValue):
(WTF::HashTraits<ASCIILiteral>::constructDeletedValue):
(WTF::HashTraits<ASCIILiteral>::isDeletedValue):
* Source/WTF/wtf/text/ASCIILiteral.h:
(WTF::ASCIILiteral::hash const):
(WTF::ASCIILiteralHash::hash):
(WTF::ASCIILiteralHash::equal):
* Source/WebCore/PAL/pal/text/TextCodec.h:
* Source/WebCore/PAL/pal/text/TextCodecCJK.cpp:
(PAL::TextCodecCJK::registerEncodingNames):
(PAL::TextCodecCJK::registerCodecs):
* Source/WebCore/PAL/pal/text/TextCodecICU.cpp:
(PAL::TextCodecICU::registerCodecs):
(PAL::TextCodecICU::TextCodecICU):
* Source/WebCore/PAL/pal/text/TextCodecICU.h:
* Source/WebCore/PAL/pal/text/TextCodecLatin1.cpp:
(PAL::TextCodecLatin1::registerEncodingNames):
(PAL::TextCodecLatin1::registerCodecs):
* Source/WebCore/PAL/pal/text/TextCodecReplacement.cpp:
(PAL::TextCodecReplacement::registerEncodingNames):
(PAL::TextCodecReplacement::registerCodecs):
* Source/WebCore/PAL/pal/text/TextCodecSingleByte.cpp:
(PAL::TextCodecSingleByte::registerEncodingNames):
(PAL::TextCodecSingleByte::registerCodecs):
* Source/WebCore/PAL/pal/text/TextCodecUTF16.cpp:
(PAL::TextCodecUTF16::registerEncodingNames):
(PAL::TextCodecUTF16::registerCodecs):
* Source/WebCore/PAL/pal/text/TextCodecUTF8.cpp:
(PAL::TextCodecUTF8::registerEncodingNames):
(PAL::TextCodecUTF8::registerCodecs):
* Source/WebCore/PAL/pal/text/TextCodecUserDefined.cpp:
(PAL::TextCodecUserDefined::registerEncodingNames):
(PAL::TextCodecUserDefined::registerCodecs):
* Source/WebCore/PAL/pal/text/TextEncoding.cpp:
(PAL::TextEncoding::decode const):
(PAL::TextEncoding::encode const):
(PAL::TextEncoding::domName const):
(PAL::TextEncoding::usesVisualOrdering const):
* Source/WebCore/PAL/pal/text/TextEncoding.h:
(PAL::TextEncoding::isValid const):
(PAL::TextEncoding::name const):
* Source/WebCore/PAL/pal/text/TextEncodingRegistry.cpp:
(PAL::TextEncodingNameHash::hash):
(PAL::HashTranslatorTextEncodingName::hash):
(PAL::HashTranslatorTextEncodingName::equal):
(PAL::isUndesiredAlias):
(PAL::WTF_REQUIRES_LOCK):
(PAL::isJapaneseEncoding):
(PAL::shouldShowBackslashAsCurrencySymbolIn):
(PAL::newTextCodec):
(PAL::atomCanonicalTextEncodingName):
* Source/WebCore/PAL/pal/text/TextEncodingRegistry.h:
* Source/WebCore/dom/DocumentInlines.h:
(WebCore::Document::encoding const):

Canonical link: https://commits.webkit.org/270903@main
